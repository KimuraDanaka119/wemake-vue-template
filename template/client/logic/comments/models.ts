// This is a definition of domain model for `comments` context.

import * as ts from 'io-ts'

/**
 * Runtime type, that can be used for schema validation.
 *
 * We call them "models", because they are exchanged between client and server.
 * @see https://github.com/gcanti/io-ts
 * @see https://github.com/aeirola/io-ts-promise
 */
export const RawComment = ts.type({
  'id': ts.number,
  'body': ts.string,
  'email': ts.string,
})

// Static TypeScript type, that can be used as a regular `type`:
export type RawCommentType = ts.TypeOf<typeof RawComment>
