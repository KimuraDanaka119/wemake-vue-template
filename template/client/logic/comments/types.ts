import { RawCommentType } from '~/logic/comments/models'

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number
}

export interface CommentPayloadType {
  commentId: number
  delta: number
}
