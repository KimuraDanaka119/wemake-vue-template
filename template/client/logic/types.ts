import { CommentType } from '~/logic/comments/types'

// TODO: document

export interface StateType {
  // Module:
  comments: {
    // Module's state:
    comments: CommentType[]
  }
}
