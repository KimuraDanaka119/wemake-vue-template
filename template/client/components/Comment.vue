<template>
  <div
    class="comment__component"
  >
    <p class="comment-author">{{ comment.email }}</p>
    <p class="comment-body">{{ comment.body }}</p>

    <div class="comment-rating">
      <button
        @click="changeRating(comment.id, 1)"
        class="comment-rating-up">+</button>

      <span class="comment-rating-value">{{ comment.rating }}</span>

      <button
        @click="changeRating(comment.id, -1)"
        class="comment-rating-down">-</button>
    </div>
  </div>
</template>

<script>
// @flow

import * as mutationTypes from '~/types/mutations'

export default {
  props: {
    comment: {
      type: Object,
      required: true
    }
  },

  methods: {
    changeRating (commentId: number, delta: number) {
      this.$store.commit(mutationTypes.UPDATE_RATING, { commentId, delta })
    }
  }
}
</script>

<style lang="scss" scoped>
.comment__component {
  background: white;
  max-width: 15.6rem;
  border: 0.12rem solid;
  padding: 0.74rem;
  margin: 0.5rem;
  position: relative;

  .comment-author {
    padding-bottom: 0.74rem;
    font-weight: bold;
    border-bottom: 0.12rem solid;
  }

  .comment-rating {
    position: absolute;
    bottom: 0.3rem;
    left: 50%;
    transform: translateX(-50%);

    button {
      margin: 0 0.5rem;
      border: none;
      outline: none;
      cursor: pointer;
      width: 2rem;

      &.comment-rating-up {
        background: #a2dca2;
      }

      &.comment-rating-down {
        background: #efbcd1;
      }
    }
  }
}
</style>
