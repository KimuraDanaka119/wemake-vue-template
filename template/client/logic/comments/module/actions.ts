import { ActionTree, ActionContext, Store } from 'vuex'

import comments from '~/logic/comments/services/api'
import * as reducers from '~/logic/comments/module/reducers'
import { StateType } from '~/logic/comments/types'

export const actions: ActionTree<StateType, StateType> = {
  async fetchComments (
    this: Store<StateType>,
    { commit }: ActionContext<StateType, StateType>,
  ) {
    const commentsList = await comments.fetchComments(this.$axios)
    commit(reducers.SET_COMMENTS, commentsList)
    return commentsList
  },
}
