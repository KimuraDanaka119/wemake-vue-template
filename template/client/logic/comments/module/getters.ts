import { GetterTree } from 'vuex'

import { StateType } from '~/logic/comments/types'

export const getters: GetterTree<StateType, StateType> = {
  hasComments (state: StateType): boolean {
    return Boolean(state.comments && state.comments.length > 0)
  },
}
