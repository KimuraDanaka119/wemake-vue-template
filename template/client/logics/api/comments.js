// @flow

import to from 'await-to-js'

import type { Axios, $AxiosError, $AxiosXHR } from 'axios'
import type { RawCommentType } from '~/types'

/**
* Throws error when there is one.
*
* @param error - Error instance to be thrown.
*/
function maybeHandleError <T> (error: ?$AxiosError<T>): void {
  if (error) {
    throw new Error(error)
  }
}

export default {
  /**
  * Fetches comments from the remote API.
  *
  * @param app - Injected Vue instance to make API calls with correct `$axios`.
  * @param app.$axios - Slightly modified `Axios` instance.
  * @throws { $AxiosError } If there is one.
  * @returns Parsed response data.
  */
  async fetchComments (
    { $axios }: { $axios: Axios }
  ): Promise<Array<RawCommentType>> {
    // Note, that $axios has some custom methods, that are not used on purpose
    // https://github.com/nuxt-community/axios-module#-features
    const [error, response]: [
      ?$AxiosError<empty>,
      $AxiosXHR<Array<RawCommentType>>
    ] = await to($axios.get('comments'))

    // Uncomment next line to see typing error:
    // console.log(response.data[0].fake)

    maybeHandleError(error)
    return response.data
  }
}
