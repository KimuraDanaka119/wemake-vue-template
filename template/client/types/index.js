// @flow

export type RawComment = {
  id: number,
  body: string,
  email: string,
};

// We add rating only on the client (for demo purposes)
export type Comment = RawComment & {
  rating: number
};

export type State = {
  comments: ?Array<Comment>
};
