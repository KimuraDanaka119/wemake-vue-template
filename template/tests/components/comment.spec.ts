import { mount, createLocalVue } from '@vue/test-utils'
import { Store } from 'vuex'

import Comment from '~/components/Comment.vue'
import { CommentType, StateType } from '~/logic/comments/types'
import * as reducers from '~/logic/comments/module/reducers'

import { storeFactory, commentFactory } from '@/fixtures/vuex'

const localVue = createLocalVue()

describe('unit tests for Comment component', () => {
  let comment: CommentType
  let store: Store<StateType>

  beforeEach(() => {
    comment = commentFactory.build({ 'rating': 0 })
    store = storeFactory.build(
      { 'state': { 'comments': [comment] } },
      { localVue }
    )
  })

  test('should have valid props', () => {
    expect.hasAssertions()

    expect(Comment).toBeValidProps({ comment })
  })

  test('should have two buttons', () => {
    expect.hasAssertions()

    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    expect(wrapper.isVueInstance()).toBe(true)
    expect(wrapper.is(Comment)).toBe(true)
    expect(wrapper.findAll('button')).toHaveLength(2)
  })

  test('should have correct values', () => {
    expect.hasAssertions()

    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    expect(wrapper.find('.body').text().trim()).toStrictEqual(comment.body)
    expect(wrapper.find('.author').text().trim()).toStrictEqual(comment.email)
    expect(wrapper.find('.number').text().trim())
      .toStrictEqual(comment.rating.toString())
  })

  test.each([
    { 'delta': 1, 'styleName': 'comment-positive' },
    { 'delta': -1, 'styleName': 'comment-negative' },
  ])('should change rating: %p', ({ delta, styleName }) => {
    expect.hasAssertions()

    const newRating = comment.rating + delta
    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    wrapper.vm.$store.commit(reducers.UPDATE_RATING, {
      'commentId': comment.id,
      delta,
    })

    expect(wrapper.vm.$store.state.comments[0].rating).toStrictEqual(newRating)
    expect(wrapper.props().comment.rating).toStrictEqual(newRating)
    expect(wrapper.classes()).toContain(wrapper.vm.$style[styleName])
  })
})

describe('snapshot tests for Comment component', () => {
  let comment: CommentType
  let store: Store<StateType>

  beforeAll(() => {
    // We need a seed here to be consistent for snapshot testing:
    comment = commentFactory.build({}, { 'seed': 8874 })
    store = storeFactory.build(
      { 'state': { 'comments': [comment] } },
      { localVue }
    )
  })

  test('should match the snapshot', () => {
    expect.hasAssertions()
    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
