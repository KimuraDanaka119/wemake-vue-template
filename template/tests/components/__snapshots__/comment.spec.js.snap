// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit tests for Comment component should match the snapshot 1`] = `
"<div class=\\"commentComponent\\"><p class=\\"author\\">test@mail.com</p> <p class=\\"body\\">some text</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">0</span> <button class=\\"ratingDown\\">
      -
    </button></div></div>"
`;
